File: android-sdk/src/main/java/com/bitmark/sdk/features/Bitmark.java
/**
 * SPDX-License-Identifier: ISC
 * Copyright © 2014-2019 Bitmark. All rights reserved.
 * Use of this source code is governed by an ISC
 * license that can be found in the LICENSE file.
 */
package com.bitmark.sdk.features;

import com.bitmark.apiservice.ApiService;
import com.bitmark.apiservice.params.*;
import com.bitmark.apiservice.params.query.BitmarkQueryBuilder;
import com.bitmark.apiservice.response.GetBitmarkResponse;
import com.bitmark.apiservice.response.GetBitmarksResponse;
import com.bitmark.apiservice.utils.Pair;
import com.bitmark.apiservice.utils.callback.Callback1;
import com.bitmark.apiservice.utils.record.BitmarkRecord;
import com.bitmark.apiservice.utils.record.ShareGrantRecord;
import com.bitmark.apiservice.utils.record.ShareRecord;

import java.util.List;

import static com.bitmark.sdk.utils.CommonUtils.wrapCallbackOnMain;

public class Bitmark {

    public static void issue(
            IssuanceParams params,
            Callback1<List<BitmarkRecord>> callback
    ) {
        ApiService.getInstance()
                .issueBitmark(params, wrapCallbackOnMain(callback));
    }

    public static void transfer(
            TransferParams params,
            Callback1<String> callback
    ) {
        ApiService.getInstance()
                .transferBitmark(params, wrapCallbackOnMain(callback));
    }

    public static void offer(
            TransferOfferParams params,
            Callback1<String> callback
    ) {
        ApiService.getInstance()
                .offerBitmark(params, wrapCallbackOnMain(callback));
    }

    public static void respond(
            TransferResponseParams params,
            Callback1<String> callback
    ) {
        ApiService.getInstance()
                .respondBitmarkOffer(params, wrapCallbackOnMain(callback));
    }

    public static void get(
            String bitmarkId, boolean loadAsset,
            Callback1<GetBitmarkResponse> callback
    ) {
        ApiService.getInstance()
                .getBitmark(bitmarkId, loadAsset, wrapCallbackOnMain(callback));
    }

    public static void get(
            String bitmarkId,
            Callback1<GetBitmarkResponse> callback
    ) {
        get(bitmarkId, false, callback);
    }

    public static void getWithAsset(
            String bitmarkId,
            Callback1<GetBitmarkResponse> callback
    ) {
        get(bitmarkId, true, callback);
    }

    public static void list(
            BitmarkQueryBuilder builder,
            Callback1<GetBitmarksResponse> callback
    ) {
        ApiService.getInstance()
                .listBitmarks(builder.build(), wrapCallbackOnMain(callback));
    }

    public static void createShare(
            ShareParams params,
            Callback1<Pair<String, String>> callback
    ) {
        ApiService.getInstance().createShare(params, callback);
    }

    public static void grantShare(
            ShareGrantingParams params,
            Callback1<String> callback
    ) {
        ApiService.getInstance().grantShare(params, callback);
    }

    public static void respondShareOffer(
            GrantResponseParams params,
            Callback1<String> callback
    ) {
        ApiService.getInstance().respondShareOffer(params, callback);
    }

    public static void getShare(
            String shareId,
            Callback1<ShareRecord> callback
    ) {
        ApiService.getInstance().getShare(shareId, callback);
    }

    public static void listShares(
            String owner,
            Callback1<List<ShareRecord>> callback
    ) {
        ApiService.getInstance().listShares(owner, callback);
    }

    public static void listShareOffer(
            String from, String to,
            Callback1<List<ShareGrantRecord>> callback
    ) {
        ApiService.getInstance().listShareOffer(from, to, callback);
    }

}

File: android-ktx/src/main/java/com/bitmark/ktx/api/Bitmark.kt
/**
 * SPDX-License-Identifier: ISC
 * Copyright © 2014-2019 Bitmark. All rights reserved.
 * Use of this source code is governed by an ISC
 * license that can be found in the LICENSE file.
 */
package com.bitmark.ktx.api

import android.util.Pair
import com.bitmark.apiservice.params.*
import com.bitmark.apiservice.params.query.BitmarkQueryBuilder
import com.bitmark.apiservice.response.GetBitmarkResponse
import com.bitmark.apiservice.response.GetBitmarksResponse
import com.bitmark.apiservice.utils.callback.Callback1
import com.bitmark.apiservice.utils.record.ShareGrantRecord
import com.bitmark.apiservice.utils.record.ShareRecord
import io.reactivex.Single

class Bitmark {

    companion object {

        fun issue(params: IssuanceParams) = Single.create<List<String>> { emt ->
            com.bitmark.sdk.features.Bitmark.issue(params, object : Callback1<List<String>> {
                override fun onSuccess(data: List<String>?) {
                    emt.onSuccess(data!!)
                }

                override fun onError(throwable: Throwable?) {
                    emt.onError(throwable!!)
                }
            })
        }

        fun transfer(params: TransferParams) = Single.create<String> { emt ->
            com.bitmark.sdk.features.Bitmark.transfer(params, object : Callback1<String> {
                override fun onSuccess(data: String?) {
                    emt.onSuccess(data!!)
                }

                override fun onError(throwable: Throwable?) {
                    emt.onError(throwable!!)
                }

            })
        }

        fun offer(params: TransferOfferParams) = Single.create<String> { emt ->
            com.bitmark.sdk.features.Bitmark.offer(params, object : Callback1<String> {
                override fun onSuccess(data: String?) {
                    emt.onSuccess(data!!)
                }

                override fun onError(throwable: Throwable?) {
                    emt.onError(throwable!!)
                }

            })
        }

        fun respond(params: TransferResponseParams) = Single.create<String> { emt ->
            com.bitmark.sdk.features.Bitmark.respond(params, object : Callback1<String> {
                override fun onSuccess(data: String?) {
                    emt.onSuccess(data!!)
                }

                override fun onError(throwable: Throwable?) {
                    emt.onError(throwable!!)
                }

            })
        }

        fun get(bitmarkId: String, loadAsset: Boolean = false) =
            Single.create<GetBitmarkResponse> { emt ->
                com.bitmark.sdk.features.Bitmark.get(bitmarkId,
                                                     loadAsset,
                                                     object : Callback1<GetBitmarkResponse> {
                                                         override fun onSuccess(data: GetBitmarkResponse?) {
                                                             emt.onSuccess(data!!)
                                                         }

                                                         override fun onError(throwable: Throwable?) {
                                                             emt.onError(throwable!!)
                                                         }
                                                     })
            }

        fun list(builder: BitmarkQueryBuilder) = Single.create<GetBitmarksResponse> { emt ->
            com.bitmark.sdk.features.Bitmark.list(builder, object : Callback1<GetBitmarksResponse> {
                override fun onSuccess(data: GetBitmarksResponse?) {
                    emt.onSuccess(data!!)
                }

                override fun onError(throwable: Throwable?) {
                    emt.onError(throwable!!)
                }
            })
        }

        fun createShare(params: ShareParams) = Single.create<Pair<String, String>> { emt ->
            com.bitmark.sdk.features.Bitmark.createShare(params, object :
                Callback1<com.bitmark.apiservice.utils.Pair<String, String>> {
                override fun onSuccess(data: com.bitmark.apiservice.utils.Pair<String, String>?) {
                    emt.onSuccess(Pair(data!!.first(), data.second()))
                }

                override fun onError(throwable: Throwable?) {
                    emt.onError(throwable!!)
                }

            })
        }

        fun grantShare(params: ShareGrantingParams) = Single.create<String> { emt ->
            com.bitmark.sdk.features.Bitmark.grantShare(params, object : Callback1<String> {
                override fun onSuccess(data: String?) {
                    emt.onSuccess(data!!)
                }

                override fun onError(throwable: Throwable?) {
                    emt.onError(throwable!!)
                }

            })
        }

        fun respondShareOffer(params: GrantResponseParams) = Single.create<String> { emt ->
            com.bitmark.sdk.features.Bitmark.respondShareOffer(params, object : Callback1<String> {
                override fun onSuccess(data: String?) {
                    emt.onSuccess(data!!)
                }

                override fun onError(throwable: Throwable?) {
                    emt.onError(throwable!!)
                }

            })
        }

        fun getShare(shareId: String) = Single.create<ShareRecord> { emt ->
            com.bitmark.sdk.features.Bitmark.getShare(shareId, object : Callback1<ShareRecord> {
                override fun onSuccess(data: ShareRecord?) {
                    emt.onSuccess(data!!)
                }

                override fun onError(throwable: Throwable?) {
                    emt.onError(throwable!!)
                }
            })
        }

        fun listShares(owner: String) = Single.create<List<ShareRecord>> { emt ->
            com.bitmark.sdk.features.Bitmark.listShares(owner,
                                                        object : Callback1<List<ShareRecord>> {
                                                            override fun onSuccess(data: List<ShareRecord>?) {
                                                                emt.onSuccess(data!!)
                                                            }

                                                            override fun onError(throwable: Throwable?) {
                                                                emt.onError(throwable!!)
                                                            }

                                                        })
        }

        fun listShareOffer(from: String, to: String) =
            Single.create<List<ShareGrantRecord>> { emt ->
                com.bitmark.sdk.features.Bitmark.listShareOffer(from,
                                                                to,
                                                                object : Callback1<List<ShareGrantRecord>> {
                                                                    override fun onSuccess(data: List<ShareGrantRecord>?) {
                                                                        emt.onSuccess(data!!)
                                                                    }

                                                                    override fun onError(throwable: Throwable?) {
                                                                        emt.onError(throwable!!)
                                                                    }

                                                                })
            }

    }
}